File: Sources/CaptionLab/LocalASR.swift
import AVFoundation
import Foundation

/// Which "second listener" the retranscribe stage (stage 5) uses to re-hear a suspect span.
/// `.gemini` is the cloud path (unchanged default); `.localASR` runs an OFFLINE local Whisper model via the
/// sidecar below — free, no API, and the MODEL is a knob (a general multilingual Whisper by default, a
/// locale-tuned checkpoint if you point it at one). The stage's map-agreement gate and energy-peak
/// re-timing are identical either way, so swapping the listener never changes the mechanism.
enum RefineBackend: String, Sendable, CaseIterable {
    case gemini, localASR
    var label: String { self == .gemini ? "Gemini (cloud)" : "Local ASR (offline)" }
}

enum LocalASRError: LocalizedError {
    case notSetUp
    case exportFailed(String)
    case process(String)
    case decode(String)
    var errorDescription: String? {
        switch self {
        case .notSetUp: "Local ASR backend not set up — run ./setup.sh --asr (creates .venv + installs mlx-whisper)."
        case .exportFailed(let r): "Audio export for local ASR failed: \(r)"
        case .process(let r): "Local ASR process error: \(r)"
        case .decode(let r): "Could not decode local ASR output: \(r)"
        }
    }
}

/// Shells out to the Python local-ASR sidecar (`asr/local_asr.py`) running in the repo's `.venv`.
/// A GENERAL, offline "second ear": transcribes an audio span with a local Whisper-family model (MLX). It
/// is deliberately NOT tied to any one language — the model id comes from `$CAPTIONLAB_ASR_MODEL` (default a
/// general multilingual Whisper), so a Taiwan user can point it at a zh-tuned checkpoint without the pipeline
/// itself hardcoding a locale. Mirrors QwenAligner's sidecar plumbing (shared `.venv`, serial subprocess).
enum LocalASR {
    static func pythonPath() -> String {
        if let p = ProcessInfo.processInfo.environment["CAPTIONLAB_PYTHON"], !p.isEmpty { return p }
        return FileManager.default.currentDirectoryPath + "/.venv/bin/python"
    }
    static func scriptPath() -> String {
        if let p = ProcessInfo.processInfo.environment["CAPTIONLAB_ASR_SCRIPT"], !p.isEmpty { return p }
        return FileManager.default.currentDirectoryPath + "/asr/local_asr.py"
    }

    /// True when the venv python and sidecar script are both present (else the GUI shows a setup hint).
    static func isAvailable() -> Bool {
        let fm = FileManager.default
        return fm.isExecutableFile(atPath: pythonPath()) && fm.fileExists(atPath: scriptPath())
    }

    /// bcp47 → the ISO-639-1 code Whisper expects (nil to let the model auto-detect, which also lets a
    /// code-switching clip fall back gracefully).
    static func mapLanguage(_ bcp47: String?) -> String? {
        let code = (bcp47 ?? "").lowercased()
        if code.hasPrefix("zh") || code.hasPrefix("yue") { return "zh" }
        if code.hasPrefix("en") { return "en" }
        if code.hasPrefix("ja") { return "ja" }
        if code.hasPrefix("ko") { return "ko" }
        if code.hasPrefix("es") { return "es" }
        if code.hasPrefix("fr") { return "fr" }
        if code.hasPrefix("de") { return "de" }
        return nil
    }

    /// Re-transcribe the source-time span `[start,end]` of `url` with the local model. `biasHint` (e.g. the
    /// content map's dialogue for this window) is passed as Whisper's initial-prompt to nudge proper nouns.
    /// Returns the verbatim text; timing is assigned by the caller on energy peaks (same as the Gemini path).
    static func transcribeSpan(url: URL, start: Double, end: Double,
                               language: String?, biasHint: String? = nil) async throws -> String {
        guard isAvailable() else { throw LocalASRError.notSetUp }
        guard end > start else { return "" }
        let wav = try await exportWAV(from: url, range: start...end)
        defer { try? FileManager.default.removeItem(at: wav) }

        var args = [scriptPath(), "--audio", wav.path]
        if let lang = mapLanguage(language) { args += ["--language", lang] }
        if let hint = biasHint?.trimmingCharacters(in: .whitespacesAndNewlines), !hint.isEmpty {
            args += ["--prompt", String(hint.prefix(220))]
        }
        let data = try await runProcess(pythonPath(), args)
        return try decodeText(data)
    }

    // MARK: - Audio export (16 kHz mono WAV, reusing AudioTrackReader) — same as QwenAligner.exportWAV

    private static func exportWAV(from url: URL, range: ClosedRange<Double>?) async throws -> URL {
        let out = FileManager.default.temporaryDirectory.appendingPathComponent("captionlab-asr-\(UUID().uuidString.prefix(8)).wav")
        var file: AVAudioFile?
        do {
            try await AudioTrackReader.read(from: url, outputSettings: [
                AVFormatIDKey: kAudioFormatLinearPCM,
                AVSampleRateKey: 16_000,
                AVNumberOfChannelsKey: 1,
                AVLinearPCMBitDepthKey: 16,
                AVLinearPCMIsFloatKey: false,
                AVLinearPCMIsBigEndianKey: false,
                AVLinearPCMIsNonInterleaved: false,
            ], range: range) { pcm in
                if file == nil {
                    file = try AVAudioFile(forWriting: out, settings: pcm.format.settings,
                                           commonFormat: pcm.format.commonFormat, interleaved: pcm.format.isInterleaved)
                }
                try file?.write(from: pcm)
            }
        } catch { throw LocalASRError.exportFailed(error.localizedDescription) }
        guard file != nil else { throw LocalASRError.exportFailed("no audio track in \(url.lastPathComponent)") }
        return out
    }

    // MARK: - Subprocess (blocking work kept off the calling actor)

    // Serial: only ONE local-ASR subprocess runs at a time so peak RAM stays ~one model, and it never
    // contends with the Qwen aligner for the GPU (each has its own serial queue; both are heavy MLX loads).
    private static let asrQueue = DispatchQueue(label: "io.captionlab.local-asr")

    private static func runProcess(_ exe: String, _ args: [String]) async throws -> Data {
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Data, Error>) in
            asrQueue.async {
                let p = Process()
                p.executableURL = URL(fileURLWithPath: exe)
                p.arguments = args
                let outPipe = Pipe(), errPipe = Pipe()
                p.standardOutput = outPipe
                p.standardError = errPipe
                do { try p.run() } catch {
                    cont.resume(throwing: LocalASRError.process("could not launch \(exe): \(error.localizedDescription)")); return
                }
                let out = outPipe.fileHandleForReading.readDataToEndOfFile()
                let err = errPipe.fileHandleForReading.readDataToEndOfFile()
                p.waitUntilExit()
                if p.terminationStatus != 0 {
                    let msg = String(data: err, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                    cont.resume(throwing: LocalASRError.process("exit \(p.terminationStatus) — \(msg.suffix(300))"))
                } else {
                    cont.resume(returning: out)
                }
            }
        }
    }

    private struct Payload: Decodable { let text: String }

    private static func decodeText(_ data: Data) throws -> String {
        guard !data.isEmpty else { throw LocalASRError.decode("empty output") }
        do {
            return try JSONDecoder().decode(Payload.self, from: data).text.trimmingCharacters(in: .whitespacesAndNewlines)
        } catch {
            let raw = String(data: data, encoding: .utf8) ?? ""
            throw LocalASRError.decode("\(error.localizedDescription) — got: \(raw.prefix(200))")
        }
    }
}

File: asr/local_asr.py
#!/usr/bin/env python3
"""CaptionLab local ASR sidecar — a general, OFFLINE "second ear" (mlx-whisper, MLX only).

Transcribes an audio file with a LOCAL Whisper-family model on Apple Silicon and prints the verbatim
result as a normalized JSON object to STDOUT — the stable contract the Swift side depends on:

    {"text": "<verbatim transcript>", "language": "<detected/given>",
     "segments": [{"text": "<...>", "start": <seconds float>, "end": <seconds float>}, ...]}

The MODEL IS A KNOB, not a hardcode: --model / $CAPTIONLAB_ASR_MODEL selects it and DEFAULTS to a general
multilingual Whisper, so nothing in the pipeline is bound to one language. Point it at a locale-tuned
checkpoint when you want that — e.g. for Taiwan Mandarin + code-switching:

    CAPTIONLAB_ASR_MODEL=mlx-community/whisper-large-v3-mlx   # general default (works out of the box)
    CAPTIONLAB_ASR_MODEL=<your-mlx-converted-Breeze-ASR-25>   # opt-in, locale-tuned (needs MLX weights)

Usage:
    python local_asr.py --audio path.wav [--language zh] [--model <id>] [--prompt "proper nouns…"]

Only STDOUT carries the JSON. Progress bars / logs from mlx-whisper go to STDERR. On failure this exits
non-zero and writes a one-line JSON error object to STDERR (STDOUT stays empty), so the caller never has
to parse mixed output.
"""
import argparse
import json
import os
import sys

DEFAULT_MODEL = os.environ.get("CAPTIONLAB_ASR_MODEL", "mlx-community/whisper-large-v3-mlx")


def fail(msg: str, code: int = 1):
    sys.stderr.write(json.dumps({"error": msg}) + "\n")
    sys.exit(code)


def main() -> int:
    ap = argparse.ArgumentParser(description="Local Whisper ASR sidecar (mlx-whisper).")
    ap.add_argument("--audio", required=True, help="Path to an audio file (16 kHz mono WAV recommended).")
    ap.add_argument("--language", default=None, help="ISO code (zh, en, ja, …); omit to auto-detect.")
    ap.add_argument("--model", default=DEFAULT_MODEL, help="MLX Whisper repo/dir. Default: a general multilingual model.")
    ap.add_argument("--prompt", default=None, help="Optional initial-prompt bias (e.g. proper nouns from the content map).")
    args = ap.parse_args()

    if not os.path.isfile(args.audio):
        fail(f"audio file not found: {args.audio}")

    try:
        import mlx_whisper  # noqa: WPS433
    except Exception as e:  # noqa: BLE001
        fail(f"mlx-whisper not importable — run ./setup.sh --asr (or pip install mlx-whisper): {e}")

    # Verbatim, no cleanup: the caller wants what was ACTUALLY said (repeats/false starts included) so the
    # map-agreement gate downstream can judge it. word_timestamps off — timing is assigned by the Swift side
    # on energy peaks, exactly like the Gemini retranscribe path.
    kwargs = {"path_or_hf_repo": args.model, "word_timestamps": False, "condition_on_previous_text": False}
    if args.language:
        kwargs["language"] = args.language
    if args.prompt:
        kwargs["initial_prompt"] = args.prompt[:220]

    try:
        result = mlx_whisper.transcribe(args.audio, **kwargs)
    except Exception as e:  # noqa: BLE001
        fail(f"transcription failed with model '{args.model}': {e}")

    text = str(result.get("text", "")).strip()
    segments = []
    for seg in (result.get("segments") or []):
        if not isinstance(seg, dict):
            continue
        t = str(seg.get("text", "")).strip()
        if not t:
            continue
        try:
            segments.append({"text": t, "start": float(seg.get("start")), "end": float(seg.get("end"))})
        except (TypeError, ValueError):
            continue

    json.dump({"text": text, "language": result.get("language"), "segments": segments},
              sys.stdout, ensure_ascii=False)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())
